### PR TITLE
Issue691 allowed origins

### PR DESCRIPTION
**Pull request by meejah**
Fixes for #691, allowed-origin checks.

(In-progress).

### CODE REVIEW

File: autobahn/websocket/protocol.py [hunk 1]
```diff
@@ -2577,19 +2609,33 @@ def processHandshake(self):
...
+                # as "origin" to onConnect, so conceivably
+                # implementations of that could mistake "" as the same
+                # origin as another WebSocket that also just lacks on
+                # Origin header...?
+
+            # XXX to pednatically follow RFC6454, we should probably
+            # assign websocket_origin as the triple (scheme, host,
+            # port) but that would change what's in onConnect's
+            # information (for .origin). Worth breaking the API for?
+            if have_origin:
```
Comment (meejah) [inline autobahn/websocket/protocol.py]: @oberstet should we do either of the above XXX things? Or should I just change the comments?

**Comment by eastein** [comment]
IMO, treating the Origin as a triple is pretty much required by the RFC and to make things safe.

**Comment by meejah** [comment]
For the "built in" checks, I believe I am treating it as a triple that must match -- my question is more regarding the API presented via ``ConnectionRequest`` to ``onConnect``. For example, we could maintain compatibility by leaving ``.origin`` as a string with the host, and add ``.scheme`` and ``.port`` or similar -- *or* force everyone to (possibly) change existing code and make ``.origin`` a ``(scheme, host, port)`` triple.

Certainly the latter seems "more correct", but might not be worth breaking code. I guess a "third way" would be to add a new ``.websocket_origin`` as a triple, deprecate ``.origin`` (removing ``.origin`` entirely at some point).

File: autobahn/twisted/test/test_protocol.py [hunk 2]
```diff
@@ -120,3 +122,111 @@ def test_handshake_fails(self):
...
+        # we're just testing the low-level function here...
+        from autobahn.websocket.protocol import _is_same_origin
+        policy = wildcards2patterns(['example.com'])
+
+        self.assertFalse(_is_same_origin('http://example.com:8080/', 80, policy))
+        self.assertFalse(_is_same_origin('http://www.example.com/', 80, policy))
+        self.assertFalse(_is_same_origin('https://example.com:80/', 443, policy))
+        self.assertFalse(_is_same_origin('https://example.com/', 80, policy))
+        self.assertFalse(_is_same_origin('http://example.org/', 80, policy))
+        self.assertFalse(_is_same_origin('http://ietf.org/', 80, policy))
```
Comment (vetsin) [inline autobahn/twisted/test/test_protocol.py]: Additional valid URIs, though i'm unsure if these would be possible to spoof into the Origin given any browser:

* `http://example.com./`
* `//example.com/`
* `http://@example.com`

File: autobahn/twisted/test/test_protocol.py [hunk 3]
```diff
@@ -120,3 +122,111 @@ def test_handshake_fails(self):
...
+        """
+        All the example origins from RFC6454 (3.2.1)
+        """
+        # we're just testing the low-level function here...
+        from autobahn.websocket.protocol import _is_same_origin
+        policy = wildcards2patterns(['example.com'])
+
+        self.assertTrue(_is_same_origin('http://example.com/', 80, policy))
+        self.assertTrue(_is_same_origin('http://example.com:80/', 80, policy))
+        self.assertTrue(_is_same_origin('http://example.com/path/file', 80, policy))
```
Comment (vetsin) [inline autobahn/twisted/test/test_protocol.py]: * `http://example.com/;semi=true`

File: autobahn/websocket/protocol.py [hunk 4]
```diff
@@ -74,6 +75,77 @@
...
+        return 'null'
+
+    (scheme, netloc, path, _, _, _) = urllib.parse.urlparse(url)
+    host = netloc
+    if ':' in netloc:
+        host, port = netloc.split(':')
+        port = int(port)
+    else:
+        try:
+            port = {'https': 443, 'http': 80, 'file': 80}[scheme]
```
Comment (eastein) [inline autobahn/websocket/protocol.py]: does it really make sense to treat file:// as port 80? Why do that?

File: autobahn/websocket/protocol.py [hunk 5]
```diff
@@ -74,6 +75,77 @@
...
+    if websocket_origin == 'null':
+        # nothing is the same as the null origin
+        return False
+
+    if not isinstance(websocket_origin, tuple) or not len(websocket_origin) == 3:
+        raise ValueError("'websocket_origin' must be a 3-tuple")
+
+    (origin_scheme, origin_host, origin_port) = websocket_origin
+
+    # XXX maybe better to explicitly pass in desired scheme, instead
```
Comment (eastein) [inline autobahn/websocket/protocol.py]: yeah that would be better..

**Comment by meejah** [comment]
@oberstet what do you think of 671266f (comes with example code)? This allows the existing API to function while also allowing one to be explicit about ports. Perhaps I should also remove the explicit "scheme" check (in `_is_same_origin()`), too? (And rely on the reg-ex?). Currently, this will *not* allow you to accept e.g. an http origin at an https websocket endpoint (or vice-versa) -- but that probably shouldn't be allowed anyway.

**Comment by oberstet** [comment]
@meejah The pragmatic approach looks good to me. Filtering on origin in a way is snake oil anyway: a non-browser client can fake it. Could you squash the PR and merge?

**Pull request merged**